packages/eslint-plugin/src/rules/no-unnecessary-type-parameters.ts

```diff
@@ -0,0 +1,374 @@
+        'ClassDeclaration[typeParameters]',
+        'ClassExpression[typeParameters]',
+        'FunctionDeclaration[typeParameters]',
+        'FunctionExpression[typeParameters]',
+        'TSCallSignatureDeclaration[typeParameters]',
+        'TSConstructorType[typeParameters]',
+        'TSDeclareFunction[typeParameters]',
+        'TSEmptyBodyFunctionExpression[typeParameters]',
+        'TSFunctionType[typeParameters]',
+        'TSMethodSignature[typeParameters]',
+      ].join(', ')](node: TSESTree.FunctionLike): void {
+        const tsNode = parserServices.esTreeNodeToTSNodeMap.get(
+          node,
+        ) as NodeWithTypeParameters;
+
+        const checker = parserServices.program.getTypeChecker();
+        let counts: Map<ts.Identifier, number> | undefined;
+
+        for (const typeParameter of tsNode.typeParameters) {
+          const esTypeParameter =
+            parserServices.tsNodeToESTreeNodeMap.get<TSESTree.TSTypeParameter>(
+              typeParameter,
+            );
+          const scope = context.sourceCode.getScope(esTypeParameter);
+
+          // Quick path: if the type parameter is used multiple times in the AST,
+          // we don't need to dip into types to know it's repeated.
+          if (isTypeParameterRepeatedInAST(esTypeParameter, scope.references)) {
+            continue;
+          }
+
+          // For any inferred types, we have to dip into type checking.
+          counts ??= countTypeParameterUsage(checker, tsNode);
+          const identifierCounts = counts.get(typeParameter.name);
+          if (!identifierCounts || identifierCounts > 2) {
+            continue;
+          }
+
+          context.report({
+            data: {
+              name: typeParameter.name.text,
+            },
+            node: esTypeParameter,
+            messageId: 'sole',
+          });
+        }
+      },
+    };
+  },
+});
+
+function isTypeParameterRepeatedInAST(
+  node: TSESTree.TSTypeParameter,
+  references: Reference[],
+): boolean {
+  let total = 0;
+
+  for (const reference of references) {
+    // References inside the type parameter's definition don't count.
+    if (
+      reference.identifier.range[0] < node.range[1] &&
+      reference.identifier.range[1] > node.range[0]
+    ) {
+      continue;
+    }
+
+    // Neither do references that aren't to the same type parameter,
+    // namely value-land (non-type) identifiers of the type parameter's type,
+    // and references to different type parameters or values.
+    if (
+      !reference.isTypeReference ||
+      reference.identifier.name !== node.name.name
+    ) {
+      continue;
+    }
+
+    // If the type parameter is being used as a type argument, then we
+    // know the type parameter is being reused and can't be reported.
+    if (
+      reference.identifier.parent.type === AST_NODE_TYPES.TSTypeReference &&
+      reference.identifier.parent.parent.type ===
+        AST_NODE_TYPES.TSTypeParameterInstantiation &&
```

Comment:
This is a little more strict than necessary. It misses cases where the type parameter is used in a type expression that's passed as the type argument, for example `Promise<T | null>`. The slow path will still handle this correctly, but might be a nice optimization in the future.

Reply:
Great spot!